### PR TITLE
Simplify and generalize BezierSegment.

### PR DESCRIPTION
**Pull request by anntzer**
As it turns out, it is simpler to write the d-dimensional version than
the 2-dimensional special case.

I initially tried to rebase the original PR (#9686, from @ijackson) but the codebase has diverged to much for it to make sense.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/bezier.py [hunk 1]
```diff
@@ -156,24 +156,21 @@ class BezierSegment(object):
 
     def __init__(self, control_points):
         """
-        *control_points* : location of contol points. It needs have a
-         shape of n * 2, where n is the order of the bezier line. 1<=
-         n <= 3 is supported.
+        Parameters
+        ----------
+        control_points : (N, D) array
+            Location of control points.  N = 2, 3 and 4 (linear, quadratic, and
```
Comment (dstansby) [inline lib/matplotlib/bezier.py]: Should this be ``D = 2, 3, 4``?

Comment (anntzer) [reply lib/matplotlib/bezier.py]: No, it's N.

**Comment by anntzer** [comment]
edit: let's also support arbitrary *N*, while we're at it...

**Comment by anntzer** [comment]
But the patch is indirectly tested by other tests (as shown by coverage).

**Pull request merged**